"Play" Tab -> Table/Tile issues

Actual Issue:
In Tile view, checking/unchecking "Show password protected" / "show game with mods" does not work. I have to switch to Table view and then back to Tile view to apply the checkbox changes.

Annoying but not an issue:
In Table view, when there are many mods in a lobby, i cannot see all mods - it only shows e.g. "mod1, mod2 and 3 more...". i have to switch to Tile view and then hover over the lobby - very unintuitive.

Please allways post logs with such issues.
It is caused by the following exception:
```
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: Bidirectional binding failed, setting to the previous value
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:284)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:227)
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:361)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.BooleanPropertyBase.fireValueChangedEvent(BooleanPropertyBase.java:103)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:266)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:227)
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:361)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.BooleanPropertyBase.fireValueChangedEvent(BooleanPropertyBase.java:103)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.beans.property.BooleanPropertyBase.access$000(BooleanPropertyBase.java:49)
	at javafx.beans.property.BooleanPropertyBase$Listener.invalidated(BooleanPropertyBase.java:245)
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:349)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.BooleanPropertyBase.fireValueChangedEvent(BooleanPropertyBase.java:103)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:264)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:227)
	at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:361)
	at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
	at javafx.beans.property.BooleanPropertyBase.fireValueChangedEvent(BooleanPropertyBase.java:103)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.scene.control.CheckBox.setSelected(CheckBox.java:156)
	at javafx.scene.control.CheckBox.fire(CheckBox.java:236)
	at com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:182)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:96)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:89)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$353(GlassViewEventHandler.java:432)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$147(WinApplication.java:177)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: TableColumn.visible : A bound value cannot be set.
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:140)
	at com.sun.javafx.binding.BidirectionalBinding$BidirectionalBooleanBinding.changed(BidirectionalBinding.java:264)
	... 64 more
```

--- a/src/main/java/com/faforever/client/game/CustomGamesController.java
+++ b/src/main/java/com/faforever/client/game/CustomGamesController.java
@@ -83,6 +83,7 @@ public class CustomGamesController extends AbstractViewController<Node> {
   public CheckBox showPasswordProtectedGamesCheckBox;
   private final ChangeListener<Boolean> filterConditionsChangedListener = (observable, oldValue, newValue) -> updateFilteredItems();
   private GamesTilesContainerController gamesTilesContainerController;
+  private final ChangeListener<Game> gameChangeListener;
 
   public CustomGamesController(UiService uiService, GameService gameService, PreferencesService preferencesService,
                                EventBus eventBus, I18n i18n) {
@@ -91,6 +92,8 @@ public class CustomGamesController extends AbstractViewController<Node> {
     this.preferencesService = preferencesService;
     this.eventBus = eventBus;
     this.i18n = i18n;
+
+    gameChangeListener = (observable, oldValue, newValue) -> setSelectedGame(newValue);
   }
 
   public void initialize() {
@@ -130,8 +133,6 @@ public class CustomGamesController extends AbstractViewController<Node> {
     } else {
       viewToggleGroup.selectToggle(tableButton);
       tableButton.getOnAction().handle(null);
-      gamesTableController.setModsColumnVisibility(showModdedGamesCheckBox.selectedProperty().getValue());
-      gamesTableController.setPasswordProtectionColumnVisibility(showPasswordProtectedGamesCheckBox.selectedProperty().getValue());
     }
     viewToggleGroup.selectedToggleProperty().addListener((observable, oldValue, newValue) -> {
       if (newValue == null) {
@@ -167,10 +168,6 @@ public class CustomGamesController extends AbstractViewController<Node> {
     filteredItems.setPredicate(OPEN_CUSTOM_GAMES_PREDICATE.and(gameInfoBean ->
         (showPasswordProtectedGames || !gameInfoBean.getPasswordProtected())
             && (showModdedGames || gameInfoBean.getSimMods().isEmpty())));
-    if (tableButton.isSelected()) {
-      gamesTableController.setModsColumnVisibility(showModdedGamesCheckBox.selectedProperty().getValue());
-      gamesTableController.setPasswordProtectionColumnVisibility(showPasswordProtectedGamesCheckBox.selectedProperty().getValue());
-    }
   }
 
   public void onCreateGameButtonClicked() {
@@ -224,7 +221,7 @@ public class CustomGamesController extends AbstractViewController<Node> {
 
   public void onTilesButtonClicked() {
     gamesTilesContainerController = uiService.loadFxml("theme/play/games_tiles_container.fxml");
-    gamesTilesContainerController.selectedGameProperty().addListener((observable, oldValue, newValue) -> setSelectedGame(newValue));
+    JavaFxUtil.addListener(gamesTilesContainerController.selectedGameProperty(), new WeakChangeListener<>(gameChangeListener));
     chooseSortingTypeChoiceBox.getItems().clear();
 
     Platform.runLater(() -> {

--- a/src/main/java/com/faforever/client/game/GameService.java
+++ b/src/main/java/com/faforever/client/game/GameService.java
@@ -52,9 +52,9 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.InitializingBean;
 import org.springframework.context.annotation.Lazy;
 import org.springframework.stereotype.Service;
-import org.springframework.beans.factory.InitializingBean;
 
 import javax.inject.Inject;
 import java.io.IOException;
@@ -269,7 +269,6 @@ public class GameService implements InitializingBean {
   /**
    * @param path a replay file that is readable by the preferences without any further conversion
    */
-
   public void runWithReplay(Path path, @Nullable Integer replayId, String featuredMod, Integer version, Map<String, Integer> modVersions, Set<String> simMods, String mapName) {
     if (isRunning()) {
       logger.warn("Forged Alliance is already running, not starting replay");
@@ -334,7 +333,6 @@ public class GameService implements InitializingBean {
     return games;
   }
 
-
   public Game getByUid(int uid) {
     Game game = uidToGameInfoBean.get(uid);
     if (game == null) {
@@ -394,7 +392,6 @@ public class GameService implements InitializingBean {
    * Returns the preferences the player is currently in. Returns {@code null} if not in a preferences.
    */
   @Nullable
-
   public Game getCurrentGame() {
     synchronized (currentGame) {
       return currentGame.get();

--- a/src/main/java/com/faforever/client/game/GamesTableController.java
+++ b/src/main/java/com/faforever/client/game/GamesTableController.java
@@ -16,7 +16,9 @@ import javafx.beans.binding.Bindings;
 import javafx.beans.property.ObjectProperty;
 import javafx.beans.property.SimpleObjectProperty;
 import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.value.ChangeListener;
 import javafx.beans.value.ObservableValue;
+import javafx.beans.value.WeakChangeListener;
 import javafx.collections.ObservableList;
 import javafx.collections.transformation.SortedList;
 import javafx.geometry.Pos;
@@ -63,6 +65,8 @@ public class GamesTableController implements Controller<Node> {
   public TableColumn<Game, String> hostColumn;
   public TableColumn<Game, Boolean> passwordProtectionColumn;
   public TableColumn<Game, String> coopMissionName;
+  private final ChangeListener<Boolean> showModdedGamesChangedListener;
+  private final ChangeListener<Boolean> showPasswordProtectedGamesChangedListener;
 
   @Inject
   public GamesTableController(MapService mapService, JoinGameHelper joinGameHelper, I18n i18n, UiService uiService, PreferencesService preferencesService) {
@@ -73,6 +77,9 @@ public class GamesTableController implements Controller<Node> {
     this.preferencesService = preferencesService;
 
     this.selectedGame = new SimpleObjectProperty<>();
+
+    showModdedGamesChangedListener = (observable, oldValue, newValue) -> modsColumn.setVisible(newValue);
+    showPasswordProtectedGamesChangedListener = (observable, oldValue, newValue) -> passwordProtectionColumn.setVisible(newValue);
   }
 
   public ObjectProperty<Game> selectedGameProperty() {
@@ -131,14 +138,10 @@ public class GamesTableController implements Controller<Node> {
 
     gamesTable.getSelectionModel().selectedItemProperty().addListener((observable, oldValue, newValue)
         -> Platform.runLater(() -> selectedGame.set(newValue)));
-  }
 
-  public void setModsColumnVisibility(boolean isVisible) {
-    modsColumn.setVisible(isVisible);
-  }
-
-  public void setPasswordProtectionColumnVisibility(boolean isVisible) {
-    passwordProtectionColumn.setVisible(isVisible);
+    //bindings do not work as that interferes with some bidirectional bindings in the TableView itself
+    JavaFxUtil.addListener(preferencesService.getPreferences().showModdedGamesProperty(), new WeakChangeListener<>(showModdedGamesChangedListener));
+    JavaFxUtil.addListener(preferencesService.getPreferences().showPasswordProtectedGamesProperty(), new WeakChangeListener<>(showPasswordProtectedGamesChangedListener));
   }
 
   private void applyLastSorting(TableView<Game> gamesTable) {

--- a/src/main/java/com/faforever/client/game/GamesTilesContainerController.java
+++ b/src/main/java/com/faforever/client/game/GamesTilesContainerController.java
@@ -19,6 +19,7 @@ import javafx.scene.control.ComboBox;
 import javafx.scene.control.ScrollPane;
 import javafx.scene.layout.FlowPane;
 import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
 import org.springframework.beans.factory.config.ConfigurableBeanFactory;
 import org.springframework.context.annotation.Scope;
 import org.springframework.stereotype.Component;
@@ -29,6 +30,7 @@ import java.util.HashMap;
 import java.util.Locale;
 import java.util.Map;
 
+@Slf4j
 @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
 @Component
 public class GamesTilesContainerController implements Controller<Node> {
@@ -36,6 +38,7 @@ public class GamesTilesContainerController implements Controller<Node> {
   private final UiService uiService;
   private final ListChangeListener<Game> gameListChangeListener;
   private final PreferencesService preferencesService;
+  private final GameService gameService;
   public FlowPane tiledFlowPane;
   public ScrollPane tiledScrollPane;
   private final ChangeListener<? super TilesSortingOrder> sortingListener;
@@ -45,9 +48,10 @@ public class GamesTilesContainerController implements Controller<Node> {
   Map<Integer, Node> uidToGameCard;
 
   @Inject
-  public GamesTilesContainerController(UiService uiService, PreferencesService preferencesService) {
+  public GamesTilesContainerController(UiService uiService, PreferencesService preferencesService, GameService gameService) {
     this.uiService = uiService;
     this.preferencesService = preferencesService;
+    this.gameService = gameService;
     selectedGame = new SimpleObjectProperty<>();
 
     sortingListener = (observable, oldValue, newValue) -> {
@@ -60,15 +64,20 @@ public class GamesTilesContainerController implements Controller<Node> {
       sortNodes();
     };
 
-    gameListChangeListener = change -> Platform.runLater(() -> {
-      synchronized (change) {
+    gameListChangeListener = change -> {
+      Platform.runLater(() -> {
         while (change.next()) {
-          change.getRemoved().forEach(gameInfoBean -> tiledFlowPane.getChildren().remove(uidToGameCard.remove(gameInfoBean.getId())));
+          change.getRemoved().forEach(gameInfoBean -> {
+            boolean remove = tiledFlowPane.getChildren().remove(uidToGameCard.remove(gameInfoBean.getId()));
+            if (!remove) {
+              log.error("Tried to remove game tile that did not exist.");
+            }
+          });
           change.getAddedSubList().forEach(GamesTilesContainerController.this::addGameCard);
+          sortNodes();
         }
-        sortNodes();
-      }
-    });
+      });
+    };
   }
 
   private void sortNodes() {
@@ -88,8 +97,12 @@ public class GamesTilesContainerController implements Controller<Node> {
   void createTiledFlowPane(ObservableList<Game> games, ComboBox<TilesSortingOrder> choseSortingTypeChoiceBox) {
     initializeChoiceBox(choseSortingTypeChoiceBox);
     uidToGameCard = new HashMap<>();
+
+    JavaFxUtil.assertApplicationThread();
+    //No lock is needed here because game updates are always done on the Application thread
     games.forEach(this::addGameCard);
     JavaFxUtil.addListener(games, new WeakListChangeListener<>(gameListChangeListener));
+
     selectFirstGame();
     sortNodes();
   }

--- a/src/test/java/com/faforever/client/game/GamesTilesContainerControllerTest.java
+++ b/src/test/java/com/faforever/client/game/GamesTilesContainerControllerTest.java
@@ -5,6 +5,7 @@ import com.faforever.client.preferences.Preferences;
 import com.faforever.client.preferences.PreferencesService;
 import com.faforever.client.test.AbstractPlainJavaFxTest;
 import com.faforever.client.theme.UiService;
+import javafx.application.Platform;
 import javafx.beans.Observable;
 import javafx.collections.FXCollections;
 import javafx.collections.ObservableList;
@@ -36,13 +37,15 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
   private UiService uiService;
   @Mock
   private PreferencesService preferencesService;
+  @Mock
+  private GameService gameService;
 
   private GamesTilesContainerController instance;
   private Preferences preferences;
 
   @Before
   public void setUp() throws Exception {
-    instance = new GamesTilesContainerController(uiService, preferencesService);
+    instance = new GamesTilesContainerController(uiService, preferencesService, gameService);
 
     when(uiService.loadFxml("theme/play/game_card.fxml")).thenReturn(gameTileController);
     preferences = new Preferences();
@@ -56,7 +59,14 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
   public void testCreateTiledFlowPaneWithEmptyList() throws Exception {
     ObservableList<Game> observableList = FXCollections.observableArrayList();
 
-    instance.createTiledFlowPane(observableList, new ComboBox<>());
+    CountDownLatch createdTiledFlowPaneCountDown = new CountDownLatch(1);
+
+    Platform.runLater(() -> {
+      instance.createTiledFlowPane(observableList, new ComboBox<>());
+      createdTiledFlowPaneCountDown.countDown();
+    });
+
+    createdTiledFlowPaneCountDown.await();
 
     assertThat(instance.tiledFlowPane.getChildren(), empty());
   }
@@ -67,8 +77,14 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
     ObservableList<Game> observableList = FXCollections.observableArrayList();
     observableList.add(new Game());
 
-    instance.createTiledFlowPane(observableList, new ComboBox<>());
+    CountDownLatch initializedCountDown = new CountDownLatch(1);
 
+    Platform.runLater(() -> {
+      instance.createTiledFlowPane(observableList, new ComboBox<>());
+      initializedCountDown.countDown();
+    });
+
+    initializedCountDown.await();
     assertThat(instance.tiledFlowPane.getChildren(), hasSize(1));
   }
 
@@ -81,8 +97,10 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
 
     ObservableList<Game> observableList = FXCollections.observableArrayList();
 
-    instance.createTiledFlowPane(observableList, new ComboBox<>());
-    observableList.add(new Game());
+    Platform.runLater(() -> {
+      instance.createTiledFlowPane(observableList, new ComboBox<>());
+      observableList.add(new Game());
+    });
 
     latch.await();
     assertThat(instance.tiledFlowPane.getChildren(), hasSize(1));
@@ -90,17 +108,19 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
 
   @Test
   public void testCreateTiledFlowPaneWithPopulatedListAndPostInstantiatedGameInfoBean() throws Exception {
-    CountDownLatch latch = new CountDownLatch(2);
+    CountDownLatch latch = new CountDownLatch(3);
     ObservableList<Node> children = instance.tiledFlowPane.getChildren();
     children.addListener((Observable observable) -> latch.countDown());
 
     doAnswer(invocation -> new Pane()).when(gameTileController).getRoot();
 
     ObservableList<Game> observableList = FXCollections.observableArrayList();
-
-    observableList.add(GameBuilder.create().defaultValues().get());
-    instance.createTiledFlowPane(observableList, new ComboBox<>());
-    observableList.add(GameBuilder.create().defaultValues().get());
+    Platform.runLater(() -> {
+      observableList.add(GameBuilder.create().defaultValues().get());
+      instance.createTiledFlowPane(observableList, new ComboBox<>());
+      observableList.add(GameBuilder.create().defaultValues().get());
+      latch.countDown();
+    });
 
     latch.await();
 
@@ -127,10 +147,15 @@ public class GamesTilesContainerControllerTest extends AbstractPlainJavaFxTest {
 
 
     observableList.addAll(game1, game2);
-
     preferences.setGameTileSortingOrder(TilesSortingOrder.PLAYER_ASC);
 
-    instance.createTiledFlowPane(observableList, new ComboBox<>());
+    CountDownLatch createdTiledFlowPaneCountDown = new CountDownLatch(1);
+    Platform.runLater(() -> {
+      instance.createTiledFlowPane(observableList, new ComboBox<>());
+      createdTiledFlowPaneCountDown.countDown();
+    });
+    createdTiledFlowPaneCountDown.await();
+
     assertEquals(instance.uidToGameCard.get(game2.getId()), instance.tiledFlowPane.getChildren().get(0));
   }
 }

